_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,350 @@
+
+The `transactional.id` is used to uniquely identify the same logical producer across process restarts.
+In contrast to the idempotent producer, a transactional producer instance will be allocated the same PID (but incremented producer epoch), as any previous instance with the same `transactional.id`. 
+Together the PID and producer epoch identify a logical producer session.
+
+> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
+> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition.
+> It was fixed by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+When starting, the client application calls `Producer.initTransactions` to initialize the producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
+Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.
+
+To guarantee message ordering, a given producer can have at most one ongoing transaction (they are executed serially).
+Brokers will prevent consumers with `isolation.level=read_committed` from advancing to offsets which include open transactions.
+Messages from aborted transactions are filtered out within the consumer, and are never observed by the client application.
+
+It is important to note that transactions are only supported inside a single Kafka cluster.
+If the transaction scope includes processors with externally observable side-effects, you would need to use additional components.
+The [transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) could help in this case.
+
+#### Transaction metadata
+
+In order to implement transactions Kafka brokers need to include extra book-keeping information in logs.
+
+Information about producers and their transactions is stored in the `__transaction_state` topic which is used by the transaction coordinator running inside a broker.
+Each transaction coordinator owns a subset of the partitions in the `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
+
+Within user logs in addition to the usual batches of data records, transaction coordinators cause partition leaders append control records (commit/abort markers) to track which batches have actually been committed and which rolled back.
+Control records are not exposed to applications by Kafka consumers, as they are an internal implementation detail of the transaction support.
```

Comment:
Well, that would not be correct, because the Kafka consumer buffers them in order to know which transactions are aborted (can't be exposed) and which are committed (can be exposed).

Reply:
Got it, thanks!